rs/https_outcalls/adapter/src/rpc_server.rs

```diff
@@ -98,8 +109,144 @@ impl CanisterHttp {
+        http_connector
+            .set_connect_timeout(Some(Duration::from_secs(self.http_connect_timeout_secs)));
+
+        if let Ok(proxy_addr) = address.parse() {
+            let client = Client::builder(TokioExecutor::new()).build::<_, Full<Bytes>>(
+                HttpsConnectorBuilder::new()
+                    .with_native_roots()
+                    .expect("Failed to set native roots")
+                    .https_only()
+                    .enable_all_versions()
+                    .wrap_connector(SocksConnector {
+                        proxy_addr,
+                        auth: None,
+                        connector: http_connector,
+                    }),
+            );
+            cache.insert(address.to_string(), client);
+            self.metrics.socks_cache_size.set(cache.len() as i64);
+        }
+    }
+
+    fn compare_results(
+        &self,
+        result: &Result<http::Response<Incoming>, String>,
+        dl_result: &Result<http::Response<Incoming>, String>,
+    ) {
+        match (result, dl_result) {
+            (Ok(_), Ok(_)) => {
+                debug!(self.logger, "SOCKS_PROXY_DL: Both requests succeeded");
+            }
+            (Err(_), Err(_)) => {
+                debug!(self.logger, "SOCKS_PROXY_DL: Both requests failed");
+            }
+            (Ok(_), Err(err)) => {
+                debug!(
+                    self.logger,
+                    "SOCKS_PROXY_DL: regular request succeeded, DL request failed with error {}",
+                    err,
+                );
+            }
+            (Err(err), Ok(_)) => {
+                debug!(
+                    self.logger,
+                    "SOCKS_PROXY_DL: DL request succeeded, regular request failed with error {}",
+                    err,
+                );
+            }
+        }
+    }
+
+    // Attempts to load the socks client from the cache. If not present, creates a new socks client and adds it to the cache.
+    fn get_socks_client(
+        &self,
+        address: &str,
+    ) -> Option<Client<HttpsConnector<SocksConnector<HttpConnector>>, OutboundRequestBody>> {
+        let cache_guard = self.cache.upgradable_read();
+
+        if let Some(client) = cache_guard.get(address) {
+            Some(client.clone())
+        } else {
+            let mut cache_guard = RwLockUpgradableReadGuard::upgrade(cache_guard);
+            self.metrics.socks_cache_misses.inc();
+            self.maybe_add_socks_client_for_address(&mut cache_guard, address);
+            match cache_guard.get(address) {
+                Some(client) => Some(client.clone()),
+                None => {
+                    debug!(
+                        self.logger,
+                        "Failed to create SOCKS client for address: {}", address
+                    );
+                    None
+                }
+            }
+        }
+    }
+
+    async fn https_outcall_socks_proxy(
+        &self,
+        socks_proxy_addrs: &[String],
```

Comment:
I believe you can move out the whole vector of socks proxies from the request in line 355 and just pass in the vec.